### ISSUE TITLE
Seed for hash functions.

### ISSUE THREAD

**Issue by yakov-olkhovskiy** [ClickHouse/ClickHouse#63482]
Allow for hash functions, implementing algorithms with seed, either additional argument or parameter to seed it.

**Use case**

```
SELECT xxHash64(42)('Hello, world!');
```

**Describe the solution you'd like**

Most obvious solution is to parametrize hash functions with seed.

**Describe alternatives you've considered**

Alternatively - create functions allowing additional argument:
```
SELECT xxHash64Seeded(42, 'Hello, world!');
```

**Additional context**

https://github.com/ClickHouse/support-escalation/issues/2042

**Comment by geetptl** [comment]
I wanna take a look at this. Any proposed approach, maintainers?

**Comment by yakov-olkhovskiy** [comment]
@geetptl parametrized version is preferred:
```
SELECT xxHash64(42)('Hello, world!');
```

**Comment by geetptl** [comment]
Are there any implementations for something similar in this codebase that I can look at?

**Comment by yakov-olkhovskiy** [comment]
@geetptl some of the aggregation function implement parameters - topK for example:
https://clickhouse.com/docs/en/sql-reference/aggregate-functions/reference/topk
though for hash functions implementation can be different and maybe complicated since they use generalized implementation.

**Comment by nikitamikhaylov** [comment]
@groeneai Implement.

**Comment by groeneai** [comment]
@nikitamikhaylov on it. One design call is needed first, because the author's preferred surface is far more expensive than it looks.

The author preferred the parametric form `xxHash64(42)('x')`. But ClickHouse has no parametric scalar functions today: the parser accepts `f(params)(args)`, yet resolution rejects non-aggregate functions with `FUNCTION_CANNOT_HAVE_PARAMETERS` (`src/Analyzer/Resolve/resolveFunction.cpp:1663`, old path `src/Interpreters/ActionsVisitor.cpp:964`). Allowing parameters on scalar functions is a core analyzer + FunctionFactory change, not a hash-function change.

The argument form is trivial and already has precedent: `sipHash64Keyed`/`sipHash128Keyed` take the key as the first argument (`src/Functions/FunctionsHashing.h`).

Two questions before I write code:
1. **Surface**: parametric `xxHash64(42)('x')` (enables parametric scalars in general, large change), or argument form such as `xxHash64Seeded(seed, ...)` / overloaded `xxHash64(seed, ...)` (small, mirrors `sipHash*Keyed`)?
2. **Scope**: which functions? Natively seedable today: `xxHash32`/`xxHash64`/`xxh3` (currently hardcode seed 0), `murmurHash2_*`/`murmurHash3_*`, `metroHash64`. cityHash/farmHash seed semantics differ.

I will start as soon as you confirm.

**Comment by rschu1ze** [comment]
@groeneai Use the argument form, not the parametric form. Also, we can't pass the seed as first argument, it must be the last argument. However, most hash functions are variable-argument functions, so that won't work as well. So we need new overloads, e.g. `metroHash64` --> `metroHash64Seed`. That is very awkward, so please add a session setting which allows users to specify a seed.

**Comment by groeneai** [comment]
@rschu1ze Thanks, going with the session setting.

Plan: new setting `hash_function_seed` (UInt64, default 0). At the default of 0 every hash output is byte-identical to today, so it is backward compatible. The seedable functions read the seed from the query context and pass it to their underlying algorithm; 0 maps to the current hardcoded seed.

Scope (functions whose algorithm runs for every input type, so the seed applies uniformly): `xxHash32`, `xxHash64`, `xxh3`, `xxh3_128`, `murmurHash2_32`, `murmurHash2_64`, `murmurHash3_32`, `murmurHash3_64`, `murmurHash3_128`, `wyHash64`. The 32-bit-seed functions use the low 32 bits.

Left out, with reasons (will note in the PR for your call):
- `metroHash64`: integer inputs take the int-hash fast path and bypass the metro algorithm, so a seed would only affect string inputs (asymmetric). Can add it with that caveat, or with explicit int-path handling, if you want.
- `gccMurmurHash` / `kafkaMurmurHash`: fixed compatibility seeds, intentionally not configurable.
- `cityHash64` / `farmHash64` / `farmFingerprint64`: no seed parameter in the algorithm.
- `sipHash*`: already seedable via the `*Keyed` variants.

PR with docs, tests, and SettingsChangesHistory shortly.

**Comment by groeneai** [comment]
PR: https://github.com/ClickHouse/ClickHouse/pull/107558

**Comment by nikitamikhaylov** [comment]
But what if you want to have several hash functions with different seeds in a query?

**Comment by groeneai** [comment]
@nikitamikhaylov You're right: a single session setting is query-global and cannot express different seeds for different hash calls in one query. That, plus a stronger objection from alexey-milovidov (the seed would have to be persisted with a table's sorting/partitioning keys when a seedable hash is used in them), is why I have closed PR #107558.

The three discussed surfaces are: parametric `xxHash64(seed)('x')` (per-call, but needs core analyzer work to allow parametric scalar functions, which are currently rejected with `FUNCTION_CANNOT_HAVE_PARAMETERS`); a positional seed argument (blocked, since most hash functions are variadic); and the session setting (query-global, with the persistence concern above). I'll hold here for your and rschu1ze's call on the surface before reimplementing.
